Wrappers/Python/cil/optimisation/algorithms/FISTA.py

```diff
@@ -70,58 +126,169 @@ def __init__(self, initial=None, f=None, g=ZeroFunction(), **kwargs):
 
-        # update_previous_solution() called after update by base class
-        #i.e current solution is now in x_old, previous solution is now in x       
 
-    def update_previous_solution(self):
-        # swap the pointers to current and previous solution
+    def update_previous_solution(self):  
+        """ Swap the pointers to current and previous solution based on the :func:`~Algorithm.update_previous_solution` of the base class :class:`Algorithm`.
+        """        
         tmp = self.x_old
         self.x_old = self.x
         self.x = tmp
 
-
     def update_objective(self):
-        self.loss.append( self.f(self.x_old) + self.g(self.x_old) )    
-    
+        """ Updates the objective
+
+        .. math:: f(x) + g(x)
+
+        """
+        self.loss.append( self.f(self.x_old) + self.g(self.x_old) )
+
+
+class FISTA(ISTA):
+
+    r"""Fast Iterative Shrinkage-Thresholding Algorithm, see :cite:`BeckTeboulle_b`, :cite:`BeckTeboulle_a`.
+
+    Fast Iterative Shrinkage-Thresholding Algorithm (FISTA)
+
+    .. math::
+
+        \begin{cases}
+            y_{k} = x_{k} - \alpha\nabla f(x_{k})  \\
+            x_{k+1} = \mathrm{prox}_{\alpha g}(y_{k})\\
+            t_{k+1} = \frac{1+\sqrt{1+ 4t_{k}^{2}}}{2}\\
+            y_{k+1} = x_{k} + \frac{t_{k}-1}{t_{k-1}}(x_{k} - x_{k-1})
+        \end{cases}
+
+    is used to solve
+
+    .. math:: \min_{x} f(x) + g(x)
+
+    where :math:`f` is differentiable, :math:`g` has a *simple* proximal operator and :math:`\alpha^{k}`
+    is the :code:`step_size` per iteration.
+
+
+    Parameters
+    ----------
+
+    initial : DataContainer
+            Starting point of the algorithm
+    f : Function
+        Differentiable function
+    g : Function
+        Convex function with *simple* proximal operator
+    step_size : positive :obj:`float`, default = None
+                Step size for the gradient step of FISTA.
+                The default :code:`step_size` is :math:`\frac{1}{L}`.
+    kwargs: Keyword arguments
+        Arguments from the base class :class:`.Algorithm`.
+
+
+    Examples
+    --------
+
+    .. math:: \underset{x}{\mathrm{argmin}}\|A x - b\|^{2}_{2}
+
+
+    >>> f = LeastSquares(A, b=b, c=0.5)
+    >>> g = ZeroFunction()
+    >>> ig = Aop.domain
+    >>> fista = FISTA(initial = ig.allocate(), f = f, g = g, max_iteration=10)
+    >>> fista.run()
 
+    See also
+    --------
+    :class:`.FISTA`
+    :class:`.GD`
+
+    """
+
+    def set_step_size(self, step_size):
+
+        """Set the default step size
+        """
+        if step_size is None:
+            if isinstance(self.f.L, Number):
+                self._step_size = 1./self.f.L
+            else:
+                raise ValueError("Function f is not differentiable")
+        else:
+            self._step_size = step_size
+
+    @property
+    def provable_convergence_condition(self):
+        return self.step_size <= 1./self.f.L
+
+    def __init__(self, initial, f, g, step_size = None, **kwargs):
```

Comment:
I suppose we could get rid of the `**kwargs` and list what's needed, also `update_objective_interval` and `max_iteration`?

Reply:
I believe it is better to act on this on another PR since we need to fix this in every of our algorithms.